working/1113 - null-asserting compound assignment/proposal.md

```diff
@@ -0,0 +1,160 @@
+because we need to insert the `!` in the middle of the expression that `++x` or `x += 1` expands to.
+
+This document proposes a new syntax, `x! += 1`, `++x!` or `x!++`, 
+which is a short-hand for an expanded version like the one above, 
+but where the *read* of the value is guarded with a non-`null` asserting check.
+
+### Composite operator specification 
+
+Currently we have a multiplicity of rules for various assignable expressions/left-hand sides (LHSs):
+
+```dart
+id
+e1.id
+e1?.id
+e1[e2]
+e1?[e2]
+super.id
+super[e2]
+ClassName.id  
+```
+
+(and we'd probably have `prefix.id` and `prefix.ClassName.id` if we didn't still pretend prefixes were objects) .
+
+Then we have operations on those LHSs:
+
+```dart
+lhs = expr
+lhs op= expr
+lhs ??= expr
+lhs ||= expr // Not yet.
+lhs &&= expr // Not yet.
+lhs++
+++lhs
+```
+
+The specification has no abstraction over LHSs, so it contains one case 
+per combination of operation and LHS shape (cf. [#228](https://github.com/dart-lang/language/issues/228)).
+
+This proposal introduces (three) more operations, since the new behavior applies equally to all existing LHSs.
+
+## Proposal
+
+We introduce new operations on LHSs of the form:
+
+```dart
+lhs! op= expr
+lhs!++
+++lhs!
+```
+
+These forms are currently not valid expressions because if `lhs` is an assignable expression,
+`lhs!` is not, and `op=` and `++` only applies to assignable expressions.
+
+### Grammar
+
+Grammatically we need to allow the new syntax. This is done by modifying the following grammar productions: 
+
+```latex
+<assignmentOperator> ::= `='
+  \alt <compoundAssignmentOperator>
+  
+<unaryExpression> ::= …
+  \alt <incrementOperator> <assignableExpression>
+
+<postfixOperator> ::= <incrementOperator>  
+```
+
+to
+
+```latex
+<assignmentOperator> ::= `='
+  \alt `!'? <compoundAssignmentOperator>
+  
+<unaryExpression> ::= …
+  \alt <incrementOperator> <assignableExpression> `!'?
+
+<postfixOperator> ::= `!'? <incrementOperator>  
+```
+
+This allows one optional `!` exactly where it affects a composite operation, and nowhere else.
+
+This grammar is not ambiguous. The `++x!` looks like it could be a problem,
+but because prefix-`++` only applies to an `<assignableExpression>`, a
+ny existing valid expression starting with `++x!` would need to be followed 
+by a selector making it into an assignable expression, like `++x!.foo`,
+and because postfix selectors bind stronger than prefix operators,
+there is no valid parse which accepts `++x!` as an unary expression
+when followed by another selector like`.foo`.
+
+### Semantics
+
+For each LHS, the semantics of an expression of the form 
+``lhs `!' op= expr`` is is specified in almost the same way
+as `lhs op= expr` except that:
+
+* After *reading* `lhs`, the operation throws if the value is `null` in the same way as `lhs!` would.
+* The static analysis uses the **NON_NULL** type of the static type of `lhs` for the operator lookup.
```

Comment:
We may want to say that if the static type of `lhs` is `T` and `T` is `T0` bounded, then we use `NON_NULL(T0)` for the operator lookup.

This will allow `++x!` where the type of `x` is `X extends int?`, where the static type `X` is the same as `NON_NULL(X)`.

Reply:
I though that **NON_NULL**(<code>*X* extends int?</code>)  would be <code>*X*&int</code> (aka <code>X & **NON_NULL**(T0)</code>. If not, I might just have to say that the static computation also uses `lhs!` where the original reads `lhs`.

It won't allow `++x!` because you can't assign `int` back to `X extends int?`.

Reply:
> I though that **NON_NULL**(`X extends int?`) would be `X&int`

Ah, that's true. So with **NonNull** (maybe it was renamed?) we actually already ensure that `X extends T` gets a similar treatment as `T` itself (as opposed to several other cases where we had to fix this recently).